tests/cpp/test_multidevice_overlap.cpp

```diff
@@ -0,0 +1,283 @@
+    // clang-format off
+        std::vector<int64_t> ta_unsharded_sizes         = {params.S, num_devices, params.M/params.S, params.K/num_devices, 1       };
+        std::vector<int64_t> ta_sizes                   = {params.S, 1          , params.M/params.S, params.K/num_devices, 1       };
+        std::vector<int64_t> tb_unsharded_sizes         = {1       , num_devices, 1                , params.K/num_devices, params.N};
+        std::vector<int64_t> tb_sizes                   = {1       , 1          , 1                , params.K/num_devices, params.N};
+        std::vector<int64_t> tc_unreduced_sizes         = {params.S, 1          , params.M/params.S, params.K/num_devices, params.N};
+        std::vector<int64_t> tc_partially_reduced_sizes = {params.S, 1          , params.M/params.S,                       params.N};
+        std::vector<int64_t> tc_sizes                   = {params.S, 1          , params.M/(params.S*num_devices)        , params.N};
+    // clang-format on
+
+    // Set up input tensors. We create the full unsharded tensors and define the
+    // actual input as the shard corresponding to the current device. Having the
+    // full unsharded input on each rank makes it possible to compute the
+    // expected result locally, hence, this way of doing is convenient for
+    // validating data correctness.
+    at::TensorOptions options =
+        at::TensorOptions().dtype(at::kFloat).device(communicator->device());
+    auto ta_unsharded = at::randn(ta_unsharded_sizes, options);
+    auto tb_unsharded = at::randn(tb_unsharded_sizes, options);
+    ta = at::empty(ta_sizes, options);
+    ta.copy_(ta_unsharded.index(
+        {at::indexing::Slice(),
+         at::indexing::Slice(my_device_index, my_device_index + 1),
+         "..."}));
+    tb = at::empty(tb_sizes, options);
+    tb.copy_(tb_unsharded.index(
+        {at::indexing::Slice(),
+         at::indexing::Slice(my_device_index, my_device_index + 1),
+         "..."}));
+
+    // We pre-allocate the output and some intermediate buffers so we do not
+    // rely on torch allocator, which do not behave well with multi-stream
+    // programming.
+    tc_unreduced = at::empty(tc_unreduced_sizes, options);
+    tc_locally_reduced = at::empty(tc_partially_reduced_sizes, options);
+    tc = at::empty(tc_sizes, options);
+
+    // compute the expected output for data correctness validation
+    auto tc_unsharded_unreduced = ta_unsharded * tb_unsharded;
+    auto tc_unsharded_expected = at::sum(tc_unsharded_unreduced, {1, 3});
+    auto tc_unsharded_expected_reshaped = at::reshape(
+        tc_unsharded_expected,
+        {params.S, num_devices, params.M / (params.S * num_devices), params.N});
+    tc_expected = tc_unsharded_expected_reshaped.index(
+        {at::indexing::Slice(),
+         at::indexing::Slice(my_device_index, my_device_index + 1),
+         "..."});
+
+    // Debug print
+    if (!communicator->deviceId() && params.debug_print) {
```

Comment:
```suggestion
    if (communicator->deviceId() == 0 && params.debug_print) {
```
Nit: I tend to avoid using boolean operations on integers directly. It may be less typing but makes the program a bit hard to follow (like 10 seconds in this case for me wondering the type of DeviceId).